src/Common/ZooKeeper/ZooKeeperImpl.cpp

```diff
@@ -449,14 +452,7 @@ void ZooKeeper::connect(
 
                 if (i != 0)
                 {
-                    std::uniform_int_distribution<UInt32> fallback_session_lifetime_distribution
-                    {
-                        args.fallback_session_lifetime.min_sec,
-                        args.fallback_session_lifetime.max_sec,
-                    };
-                    UInt32 session_lifetime_seconds = fallback_session_lifetime_distribution(thread_local_rng);
-                    client_session_deadline = clock::now() + std::chrono::seconds(session_lifetime_seconds);
-
+                    auto session_lifetime_seconds = setClientSessionDeadline(args.fallback_session_lifetime.min_sec, args.fallback_session_lifetime.max_sec);
```

Comment:
do I understand that right?
Client make a connection with limitation session life time. After that timeout the client disconnects from that node and tries to choose the better one.

Reply:
What is the value of that `fallback_session_lifetime`?

I see that would be a scenario when one keeper server is down, the clients from the same az are in the loop. They disconnect from less preferred server, trying to connect to the most preferred server, that connection takes time to establish, after that they reconnect to some unpreferred server again. I see it as no positive result, just spend time to trying make some connections.
Do we intent to address such scenario?
I see it like that. Untill client see the better node, client does not unconnect from current one. That approach is challenging. It requires some back ground activity to maintain the liveness of other servers.

Reply:
One more interesting scenario.
When local keeper is down. All clients have to choose different nonlocal servers. I see that this code would choose the one lexical lower nonlocal server.

Reply:
> ... to the most preferred server, that connection takes time to establish, after that they reconnect to some unpreferred server again.

That should not happen right?  `setClientSessionDeadline` is only invoked if we are connecting to a server not in the same AZ (less preferred). In case of trying an same-AZ keeper host, 

1. at ZooKeeper.cpp Line 252, the `nodes{node}` is initialized as a single element. So i == 0 condition still applies. We would not call  `setClientSessionDeadline` within `ZooKeeperImpl::connect` 
2. at Line 255, we found local_az == connected_host_az, return directly.

i == 0 is helpful to determine connection is optimal or not in other load balancing scenario as priority can be determined beforehand. In AZ aware load balancing case, I made it `nodes{}` is always inited with one element.

Reply:
> All clients have to choose different nonlocal servers.

Good thing is that `ShuffleHost` already addresses this. We call `randomize()` here, https://github.com/ClickHouse/ClickHouse/pull/56563/files#diff-17249c3158c8311e98425540c1d48298b695b41fadb9081cafc203a3a0edd0f9R103

And use ShuffleHost::compare for sorting, which first use the priority and then host.random value for comparision.

https://github.com/ClickHouse/ClickHouse/blob/329e7b5d0a482f4e0c18256b1bc817f9f562019f/src/Common/ZooKeeper/ZooKeeper.h#L64

Reply:
I missed that there is not just host name but a struct `struct ShuffleHost`, thanks for pointing.
I see that it does randomization for equal priorities.

Reply:
True. Though that can be mitigated by having a fallback_session_lifetime (say 1 - 3 hours?) that is greater than the time that keeper not being available. Would prefer to keep as it is. It also make things simpler.

Reply:
Do we support cases when there is several servers in one az?
If we do, it is worth to check what connections distribution would be between locals servers in case of its restarts.

Reply:
> Do we support cases when there is several servers in one az?
Not yet. Do you mean it could cause overloading the remaining keeper server? that's true. will document that caveat.

Reply:
There is no procedure how the same az connections are balancing. So yeah, I suspect the case when one server holds all the connections without rebalancing.